File: hyper_rendering/src/vulkan/pipeline/compute_pipeline.rs
/*
 * Copyright (c) 2023, SkillerRaptor
 *
 * SPDX-License-Identifier: MIT
 */

use crate::{
    error::{Error, Result},
    vulkan::{
        core::device::Device,
        pipeline::{pipeline_layout::PipelineLayout, Pipeline},
        resource::shader::Shader,
    },
};

use ash::vk;
use std::{ffi::CStr, rc::Rc};

pub(crate) struct ComputePipeline {
    handle: vk::Pipeline,

    device: Rc<Device>,
}

impl ComputePipeline {
    pub(crate) fn new(device: Rc<Device>, create_info: ComputePipelineCreateInfo) -> Result<Self> {
        let ComputePipelineCreateInfo { layout, shader } = create_info;

        let entry_name = unsafe { CStr::from_bytes_with_nul_unchecked(b"main\0") };

        let stage_create_info = vk::PipelineShaderStageCreateInfo::builder()
            .stage(vk::ShaderStageFlags::COMPUTE)
            .module(shader.handle())
            .name(entry_name);

        let create_info = vk::ComputePipelineCreateInfo::builder()
            .stage(*stage_create_info)
            .layout(layout.handle())
            .base_pipeline_handle(vk::Pipeline::null())
            .base_pipeline_index(0);

        let handle = unsafe {
            device
                .handle()
                .create_compute_pipelines(vk::PipelineCache::null(), &[*create_info], None)
                .map_err(|error| Error::VulkanCreation(error.1, "compute pipeline"))
        }?[0];

        Ok(Self { handle, device })
    }
}

impl Drop for ComputePipeline {
    fn drop(&mut self) {
        unsafe {
            self.device.handle().destroy_pipeline(self.handle, None);
        }
    }
}

impl Pipeline for ComputePipeline {
    fn handle(&self) -> vk::Pipeline {
        self.handle
    }
}

pub(crate) struct ComputePipelineCreateInfo<'a> {
    layout: &'a PipelineLayout,

    shader: Shader,
}

File: hyper_rhi/src/hyper_rhi/vulkan/compute_pipeline.cpp
/*
 * Copyright (c) 2025-present, SkillerRaptor
 *
 * SPDX-License-Identifier: MIT
 */

#include "hyper_rhi/vulkan/compute_pipeline.hpp"

#include "hyper_rhi/vulkan/graphics_device.hpp"
#include "hyper_rhi/vulkan/pipeline_layout.hpp"
#include "hyper_rhi/vulkan/shader.hpp"
#include "hyper_rhi/vulkan/utils.hpp"

namespace he
{
    VulkanComputePipeline::VulkanComputePipeline(
        VulkanGraphicsDevice &graphics_device, const ComputePipelineDescriptor &desc)
        : m_graphics_device(graphics_device)
    {
        const VulkanShader *shader = graphics_device.get_internal_state<VulkanShader>(desc.shader);

        const VkPipelineShaderStageCreateInfo pipeline_shader_stage_create_info = {
            .sType = VK_STRUCTURE_TYPE_PIPELINE_SHADER_STAGE_CREATE_INFO,
            .pNext = nullptr,
            .flags = 0,
            .stage = VK_SHADER_STAGE_COMPUTE_BIT,
            .module = shader->raw(),
            .pName = desc.shader.entry().data(),
            .pSpecializationInfo = nullptr,
        };

        const VulkanPipelineLayout *pipeline_layout
            = graphics_device.get_internal_state<VulkanPipelineLayout>(desc.layout);

        const VkComputePipelineCreateInfo compute_pipeline_create_info = {
            .sType = VK_STRUCTURE_TYPE_COMPUTE_PIPELINE_CREATE_INFO,
            .pNext = nullptr,
            .flags = 0,
            .stage = pipeline_shader_stage_create_info,
            .layout = pipeline_layout->raw(),
            .basePipelineHandle = VK_NULL_HANDLE,
            .basePipelineIndex = -1,
        };

        HE_VK_CHECK(
            vkCreateComputePipelines(
                m_graphics_device.device(), VK_NULL_HANDLE, 1, &compute_pipeline_create_info, nullptr, &m_raw),
            "Failed to create vulkan compute pipeline");
        HE_ASSERT(m_raw != VK_NULL_HANDLE);
    }

    VulkanComputePipeline::~VulkanComputePipeline() { vkDestroyPipeline(m_graphics_device.device(), m_raw, nullptr); }
} // namespace he
